pubsub.go

```diff
@@ -287,28 +353,102 @@ func (p *PubsubValueStore) Cancel(name string) (bool, error) {
 	return ok, nil
 }
 
-func (p *PubsubValueStore) handleSubscription(sub *pubsub.Subscription, key string, cancel func()) {
+func (p *PubsubValueStore) handleSubscription(sub *pubsub.Subscription, key string) {
 	defer sub.Cancel()
-	defer cancel()
 
-	for {
-		msg, err := sub.Next(p.ctx)
-		if err != nil {
-			if err != context.Canceled {
-				log.Warningf("PubsubResolve: subscription error in %s: %s", key, err.Error())
+	newMsg := make(chan []byte)
+	go func() {
+		for {
+			data, err := p.handleNewMsgs(sub, key)
+			if err != nil {
+				close(newMsg)
+				return
 			}
-			return
+			newMsg <- data
 		}
-		if p.isBetter(key, msg.GetData()) {
-			err := p.ds.Put(dshelp.NewKeyFromBinary([]byte(key)), msg.GetData())
+	}()
+
+	newPeerData := make(chan []byte)
+	go func() {
+		for {
+			data, err := p.handleNewPeer(sub, key)
+			if err != nil {
+				close(newPeerData)
+				return
+			}
+			newPeerData <- data
+		}
+	}()
+
+	for {
+		var data []byte
+		select {
+		case data = <-newMsg:
+		case data = <-newPeerData:
+		}
+
+		if isBetter, _ := p.isBetter(key, data); isBetter {
+			err := p.ds.Put(dshelp.NewKeyFromBinary([]byte(key)), data)
 			if err != nil {
 				log.Warningf("PubsubResolve: error writing update for %s: %s", key, err)
 			}
-			p.notifyWatchers(key, msg.GetData())
+			p.notifyWatchers(key, data)
 		}
 	}
 }
 
+func (p *PubsubValueStore) handleNewMsgs(sub *pubsub.Subscription, key string) ([]byte, error) {
+	msg, err := sub.Next(p.ctx)
+	if err != nil {
+		if err != context.Canceled {
+			log.Warningf("PubsubResolve: subscription error in %s: %s", key, err.Error())
+		}
+		return nil, err
+	}
+	return msg.GetData(), nil
+}
+
+func (p *PubsubValueStore) handleNewPeer(sub *pubsub.Subscription, key string) ([]byte, error) {
+	peerEvt, err := sub.NextPeerEvent(p.ctx)
+	for {
+		if err != nil {
+			if err != context.Canceled {
+				log.Warningf("PubsubNewPeer: subscription error in %s: %s", key, err.Error())
+			}
+			return nil, err
+		}
+		if peerEvt.Type == pubsub.PEER_JOIN {
+			break
+		}
+	}
+
+	peerCtx, _ := context.WithTimeout(p.ctx, time.Second*10)
+	s, err := p.host.NewStream(peerCtx, peerEvt.Peer, OnJoinProto)
```

Comment:
Hm. I kind of see why you want push. (also seeing why being able to hook into the hello message in pubsub would be nice).

Reply:
For sure, let me know if you've got any ideas. As we mentioned above, while push is more efficient the inherent ACK from query-response is nice.

Reply:
I'd still stick with a request/response. The client may, for example, have a concept of "good enough" and may want to stop querying at some point.

Reply:
Going with request/response after some discussion (https://github.com/libp2p/go-libp2p/issues/698#issuecomment-518737317)